moveit_ros/hybrid_planning/local_planner/local_planner_component/src/local_planner_component.cpp

```diff
@@ -159,8 +159,9 @@ bool LocalPlannerComponent::initialize()
         RCLCPP_INFO(LOGGER, "Received local planning goal request");
         return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
       },
-      [](const std::shared_ptr<rclcpp_action::ServerGoalHandle<moveit_msgs::action::LocalPlanner>>& /*unused*/) {
+      [this](const std::shared_ptr<rclcpp_action::ServerGoalHandle<moveit_msgs::action::LocalPlanner>>& /*unused*/) {
         RCLCPP_INFO(LOGGER, "Received request to cancel local planning goal");
+        state_ = LocalPlannerState::ABORT;
```

Comment:
This looks like communication across threads with shared memory also.  See my previous comment on that.

Reply:
Agree that this is a problem!

Reply:
I can make this variable atomic. That's a lock-free synchronization primitive. I'll push this change; let me know if you can think of a better way (hopefully not ridiculously complicated).

Reply:
Even lock-free synchronization primitives still cause points of synchronization (single-threaded sections of the code).  I would much prefer a method of communication that does not rely on synchronization of memory access between threads.

The primary reason for this is that if you make something an atomic it has to be stored in ram and access to that point in ram has to be synchronized.  This means that for the CPU to modify or read the value it has to go and get it from ram and then make sure it changed it while preventing no other CPU from accessing it at the same time.  lock-free things like atomic use built-in CPU instructions for this instead of mutexes so they are more performant but they are still creating points in the code that have to be executed in a single-threaded way.

Reply:
My proposal would be to change our mental model for how we build multi-threaded applications to one that does not use mutable state shared between threads for communication.